In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv('Resources/lending_data.csv')
# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df['loan_status']

# Separate the X variable, the features
X = lending_data_df.copy()
X.drop('loan_status', axis = 1, inplace = True)

In [4]:
# Review the y variable array
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
# If the data contains non-numerical values it needs to be normalised with the pd.get_dummies function.
print(X.dtypes)
print('==============================')
X.head()

loan_size           float64
interest_rate       float64
borrower_income       int64
debt_to_income      float64
num_of_accounts       int64
derogatory_marks      int64
total_debt            int64
dtype: object


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1)

In [7]:
# Scale the data
# It's important to scale  & transform data of varying ranges & magnitudes to remove statistically insignificant bias in the data.
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_scaler = scaler.fit(X_train)

X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [28]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_instance = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_regression_instance.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [29]:
# Make a prediction using the testing data
test_predictions = logistic_regression_instance.predict(X_test_scaled)
test_predictions[:5]

array([0, 0, 0, 0, 0], dtype=int64)

In [30]:
# Make a prediction using the training data
train_predictions = logistic_model.predict(X_train_scaled)
train_predictions[:5]

array([0, 0, 0, 0, 0], dtype=int64)

### Brief Analysis

In [31]:
# Score the data
print(f"Training Data Score: {logistic_regression_instance.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {logistic_regression_instance.score(X_test_scaled, y_test)}")

Training Data Score: 0.9942908240473243
Testing Data Score: 0.9936545604622369


In [32]:
print(f"The maximum loan size was: ${'{:,.2f}'.format(lending_data_df['loan_size'].max())}")
print(f"The average loan size was: ${'{:,.2f}'.format(np.mean(lending_data_df['loan_size']))}")
print(f"The median loan size was: ${'{:,.2f}'.format(np.median(lending_data_df['loan_size']))}")
print(f"The total number of applicants was: {len(lending_data_df)}")
print(f"The number of expected high risk applicants receiving a loan accoring to the model is: {len(lending_data_df)*(113/19384)}")
print(f"The potential cost of lending to high risk applicants is: ${'{:,.2f}'.format((len(lending_data_df)*(113/19384)) * (np.median(lending_data_df['loan_size'])))}")

The maximum loan size was: $23,800.00
The average loan size was: $9,805.56
The median loan size was: $9,500.00
The total number of applicants was: 77536
The number of expected high risk applicants receiving a loan accoring to the model is: 452.0
The potential cost of lending to high risk applicants is: $4,294,000.00


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [33]:
# Generate a confusion matrix for the testing predictions
test_confusionMatrix = confusion_matrix(y_test, test_predictions)

test_confusionMatrix_df = pd.DataFrame(test_confusionMatrix, 
index=["Actual 0", "Actual 1"], 
columns=["Predicted 0", "Predicted 1"]
)

print("Test Confusion Matrix")
display(test_confusionMatrix_df)

# Print the classification report for the testing model

print("Test Classification Report")
print(classification_report(y_test, test_predictions))

Test Confusion Matrix


Predicted 0  Predicted 1
Actual 0        18652          113
Actual 1           10          609

Test Classification Report
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.98      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



In [34]:
# Generate a confusion matrix for the training model
train_confusionMatrix = confusion_matrix(y_train, train_predictions)

train_confusionMatrix_df = pd.DataFrame(train_confusionMatrix, 
index=["Actual 0", "Actual 1"], 
columns=["Predicted 0", "Predicted 1"]
)

print("Train Confusion Matrix")
display(train_confusionMatrix_df)

# Print the classification report for the training model
print("Train Classification Report")
print(classification_report(y_train, train_predictions))

Train Confusion Matrix


Predicted 0  Predicted 1
Actual 0        55974          297
Actual 1           35         1846

Train Classification Report
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     56271
           1       0.86      0.98      0.92      1881

    accuracy                           0.99     58152
   macro avg       0.93      0.99      0.96     58152
weighted avg       0.99      0.99      0.99     58152



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model is able to predict overall to 99% accuracy. To drill down further on the healthy and high risk labels:

  - Near 100% accuracy in predicting healthy loan applicants: 
    - 18652 True positive // 113 False positive
  - Approximately 98% accuracy in predicting high-risk loan applicants:
    - 609 True Negative // 10 False Negative
  
Depending on the loan size, the estimated financial return on loan for the institution, and other factors, the decision may be made that 113 false positives may present too high a financial risk to continue with the model.

---